packages/pds/src/index.ts

```diff
@@ -6,117 +6,144 @@ import 'express-async-errors'
+    const config = new ServerConfig(opts.cfg)
+    const didResolver = new DidResolver({ plcUrl: config.didPlcUrl })
+    const auth = new ServerAuth({
+      jwtSecret: cfg.jwtSecret,
+      adminPass: cfg.adminPassword,
+      didResolver,
+    })
+
+    const messageQueue = new SqlMessageQueue('pds', db)
+    streamConsumers.listen(messageQueue, blobstore, auth, keypair)
+
+    const services = createServices(messageQueue, blobstore)
+
+    const mailTransport =
+      config.emailSmtpUrl !== undefined
+        ? createTransport(config.emailSmtpUrl)
+        : createTransport({ jsonTransport: true })
+
+    const mailer = new ServerMailer(mailTransport, config)
+
+    const app = express()
+    app.use(cors())
+    app.use(loggerMiddleware)
+
+    let imgUriEndpoint = config.imgUriEndpoint
+    if (!imgUriEndpoint) {
+      const imgProcessingCache = new BlobDiskCache(config.blobCacheLocation)
+      const imgProcessingServer = new ImageProcessingServer(
+        config.imgUriSalt,
+        config.imgUriKey,
+        blobstore,
+        imgProcessingCache,
+      )
+      app.use('/image', imgProcessingServer.app)
+      imgUriEndpoint = `${config.publicUrl}/image`
+    }
+
+    const imgUriBuilder = new ImageUriBuilder(
+      imgUriEndpoint,
       config.imgUriSalt,
       config.imgUriKey,
-      blobstore,
-      imgProcessingCache,
     )
-    app.use('/image', imgProcessingServer.app)
-    imgUriEndpoint = `${config.publicUrl}/image`
-  }
 
-  const imgUriBuilder = new ImageUriBuilder(
-    imgUriEndpoint,
-    cfg.imgUriSalt,
-    cfg.imgUriKey,
-  )
-
-  const locals: Locals = {
-    logger: httpLogger,
-    db,
-    blobstore,
-    keypair,
-    auth,
-    imgUriBuilder,
-    config,
-    mailer,
-    services,
-    messageQueue,
+    const ctx = new AppContext({
+      db,
+      blobstore,
+      keypair,
+      cfg: config,
+      auth,
+      messageQueue,
+      services,
+      mailer,
+      imgUriBuilder,
+    })
+
+    const apiServer = API(ctx, {
+      payload: {
+        jsonLimit: 100 * 1024, // 100kb
+        textLimit: 100 * 1024, // 100kb
+        blobLimit: 5 * 1024 * 1024, // 5mb
+      },
+    })
+    app.use(health.createRouter(ctx))
+    app.use(apiServer.xrpc.router)
+    app.use(error.handler)
+
+    return new PDS({
+      ctx,
+      app,
+    })
   }
 
-  app.locals = locals
+  async start(): Promise<http.Server> {
+    const server = this.app.listen(this.ctx.cfg.port)
+    this.server = server
+    return new Promise((resolve, reject) => {
+      server.on('listening', () => {
+        resolve(server)
+      })
+      server.on('error', (err) => {
+        reject(err)
+      })
+    })
+  }
 
-  app.use((req, res, next) => {
-    const reqLocals: Locals = {
-      ...locals,
-      logger: req.log, // This logger is request-specific
+  async destroy(force = false): Promise<void> {
+    await this.ctx.messageQueue.destroy()
+    if (this.server) {
+      const terminator = createHttpTerminator({
```

Comment:
I think that it's possible the terminator needs to be created prior to accepting connections, but I'm not positive— just opened an issue to ask https://github.com/gajus/http-terminator/issues/42